liquid-compiler/src/parser.rs

```diff
@@ -3,652 +3,562 @@
-pub fn consume_value_token(tokens: &mut Iter<Token>) -> Result<Token> {
-    match tokens.next() {
-        Some(t) => value_token(t.clone()),
-        None => Err(unexpected_token_error(
-            "string | number | boolean | identifier",
-            NOTHING,
-        )),
+pub struct Raw<'a> {
+    text: &'a str,
+}
+impl<'a> From<Pair<'a>> for Raw<'a> {
+    fn from(element: Pair<'a>) -> Self {
+        if element.as_rule() != Rule::Raw {
+            panic!("Only rule Raw can be converted to Raw.");
+        }
+        Raw {
+            text: element.as_str(),
+        }
     }
 }
+impl<'a> Into<&'a str> for Raw<'a> {
+    fn into(self) -> &'a str {
+        self.to_str()
+    }
+}
+impl<'a> Raw<'a> {
+    pub fn to_renderable(self) -> Box<Renderable> {
+        Box::new(Text::new(self.to_str()))
+    }
+
+    pub fn to_str(self) -> &'a str {
+        self.as_str()
+    }
 
-/// Recognises a value token, returning an error if a non-value token
-/// is presented.
-pub fn value_token(t: Token) -> Result<Token> {
-    match t {
-        v @ Token::StringLiteral(_)
-        | v @ Token::IntegerLiteral(_)
-        | v @ Token::FloatLiteral(_)
-        | v @ Token::BooleanLiteral(_)
-        | v @ Token::Identifier(_) => Ok(v),
-        x => Err(unexpected_token_error(
-            "string | number | boolean | identifier",
-            Some(&x),
-        )),
+    pub fn as_str(&self) -> &'a str {
+        self.text
     }
 }
 
-/// Describes the optional trailing part of a block split.
-pub struct BlockSplit<'a> {
-    pub delimiter: String,
-    pub args: &'a [Token],
-    pub trailing: &'a [Element],
+pub struct Tag<'a> {
+    name: &'a str,
+    tokens: Box<Iterator<Item = TagToken<'a>> + 'a>,
+    as_str: &'a str,
+}
+impl<'a> From<Pair<'a>> for Tag<'a> {
+    fn from(element: Pair<'a>) -> Self {
+        if element.as_rule() != Rule::Tag {
+            panic!("Only rule Tag can be converted to Tag.");
+        }
+        let as_str = element.as_str();
+        let mut tag = element.into_inner();
+        let name = tag
+            .next()
+            .expect("A tag starts with an identifier.")
+            .as_str();
+        let tokens = Box::new(tag.map(TagToken::from));
+
+        Tag {
+            name,
+            tokens,
+            as_str,
+        }
+    }
+}
+impl<'a> Tag<'a> {
+    pub fn name(&self) -> &str {
+        self.name
+    }
+    pub fn tokens(&mut self) -> &mut Iterator<Item = TagToken<'a>> {
+        &mut self.tokens
+    }
+    pub fn as_str(&self) -> &str {
+        self.as_str
+    }
+    pub fn parse(
+        mut self,
+        next_elements: &mut TagBlock,
+        options: &LiquidOptions,
+    ) -> Result<Box<Renderable>> {
+        let (name, tokens) = (self.name, &mut self.tokens);
+
+        let next_elements = &mut next_elements.iter;
+
+        if options.tags.contains_key(name) {
+            options.tags[name].parse(name, tokens, options)
+        } else if options.blocks.contains_key(name) {
```

Comment:
Doesn't `Tag` duplicate `parse_tag`?

Reply:
On the most part yes, and they can definitely be put together. I just hadn't thought very much on the best way to do so yet, as I was more focused on getting the if/for/case tags implementable.

I'm actually thinking about taking the idea of those structures such as `Tag` and "Expression" a bit further and do that for all 'unparsed' states of other elements of the file, replacing the current functions by associated functions to these 'unparsed' states. I think that would bring more consistency to the whole parser and also allow to make those structures public (if ever needed) without having to leak Pest's API, as well as more flexibility and type safety (because we would no longer work directly with `Pair`s, which could be anything, but with structures encapsulating them). But this is something for a future PR.